### DISCUSSION TITLE
Explanation for training/bidding needed

### DISCUSSION THREAD

**Discussion by ThePokerDude** [lorserker/ben#99] [Q&A] [answered]
Hi,
I read the training/bidding/README.md

_- KT53.KQ7.KJ876.5
and is translated to:
- 0 1 0 0 1 0 0 2 0 1 1 0 0 0 0 1 0 1 0 1 0 0 1 2 0 0 0 0 0 0 0 1
So starting from Ace of spades, and each suit ends with a value for number of small cards from 2-7._

Here everything is clear to me: SA = false, KS = true, QS = false, JS = false, TS = true, 9S = false, 8S = false, 2-7S count = 2
and then hearts, diamonds and clubs

_So starting from Ace of spades, and each suit ends with a value for number of small cards from 2-7.
In a deck there is 40 high card points (hcp), and the hcp is calculated for each hand. The values are linear normalized.
The shape of a hand is the length in the 4 suits, and again it is normalized (also known as z-score or standardization)
So we end up representing the hand as
- [ 0, 0, -1.25, -0.14, -0.14, -0.14, 0.43, **0, 1, 0, 0, 1, 0, 0, 2, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 1**]_ 

here I am lost. I marked the representation of KT53.KQ7.KJ876.5 in bold. And I understand that 0, 0, -1.25, -0.14, -0.14, -0.14, 0.43
represents the 12 HCP the 4-3-5-1 distribution but I cannot recognize what is what. Could you explain?

1st = ?
2nd = ?
3rd = ?
4th = ?
5th = ?
6th = ?
7th = ?
8-39 exact cards

_**A single bid is represented as an array with 40 elements, where only 1 item is 1, and the rest is zero** (a One-Hot array), as this will help the neural network. The values of the Hot-array is

0 = PAD_START
1 = PAD_END
2 = PASS
3 = Double
4 = Redouble
5-40 = Bids, starting from 1C and ending at 7N,_

But 0 to 40 are 41 elements and not 40. So I think 7NT could not be represented, which can get unnoticed because it' very rare.

_So we now have the input record for a bid as this

[ 0, 0, 0.5, -0.71, 0.43, -0.14, 0.43,  
   0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 2, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 2,  
   1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  
   1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,  
   1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]_

So the input record for a bid consists of the reprentation of a hand and the representation of a bid?
It seems to me as 
0, 0, 0.5, -0.71, 0.43, -0.14, 0.43, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 2, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 2
represents the east hand.
But what is the rest? It looks like 3 arrays of 40 elements for where the first element is 1.
I do not know how to interpret it: PAD_START, PAD_START, PAD_START?
Let's assume I' right and it's 41 and not 40 elements. Shouldn't it look like this:

[ 0, 0, 0.5, -0.71, 0.43, -0.14, 0.43,  
   0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 2, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 2,  
   1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0  
   0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0  
   0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

1st: east hand represtation, 2nd: PAD_START, 3rd: pass, 4th: PAD_END?

And the expected output as this:
[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
as the hand would pass (Element 3 is hot).

Why should the east hand Q7.AT43.QT3.AT63 pass with its 12 HCP?

**Comment by ThorvaldAagaard** [comment]
> 5-40 = Bids, starting from 1C and ending at 7N,_

5-40 is 36 elements, but you have only 35 bids, so index 39 is 7N (element number 40)

HCP and shape is normalized
```
        self.shapes = list(map(lambda shape: (shape - 3.25)/1.75, map(get_shape, hands)))
        self.hcp = list(map(lambda point_count: (np.array([[point_count[0]]]) - 10) / 4, map(get_hcp, hands)))
```
So -1.25, -0.14, -0.14, -0.14, 0.43 translates to 5 HCP, and I assume it should be -0.11, -0.11, -0.11, 0.43 showing  3334

So there is both an error in the number and it is not matching the rest of the hand, so I will change it to
```
 [ 0, 0, 0.5, 0.43, -0.11, 1.03, -1.29, 0, 1, 0, 0, 1, 0, 0, 2, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 1]
```

PAD_START and PAD_END are non-bids used as fillers.

When you ask the Neural net for a bid, you input the hand and the three previous bids (3*40 as a bid is 40 one-hot elements.)

So 3 times PAD_START just means that the player is first hand to bid

You are correct, that the correct output should be 1C (Element 5 hot)

I have updated the document

**Comment by ThePokerDude** [comment]
Thanks a lot, now it makes perfect sense.

_So we end up representing the hand as
[ 0, 0, 0.5, 0.43, -0.11, 1.03, -1.29, 0, 1, 0, 0, 1, 0, 0, 2, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 1]_

I'm unsure about the two leading zeros. I think they represent the vulnaribilty (NS not vulnerable and EW not vulnerable).
So I think they should be removed when explaining the hand representation and this part should be changed to:

So we end up representing the hand as
[ 0.5, 0.43, -0.11, 1.03, -1.29, 0, 1, 0, 0, 1, 0, 0, 2, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 1]



0 = PAD_START 1 = PAD_END 2 = PASS 3 = Double 4 = Redouble 5-39 = Bids, starting from 1C and ending at 7N,

**Could we add here additional explanation like?**
_An input record for a bid consists of:

- Numbers 1 and 2: vulnaribility of NS and EW respectively,
- Numbers 3 to 39 representation of the hand (1st: normalized HCP, 2nd - 5th: normalized length of spades, hearts, diamonds and clubs, 6th - 37th: exact representation of spades, hearts, diamonds and clubs)
- Numbers 40 to 79: representation of the third last bid
- Numbers 80 to 119: representation of the second last bid
- Numbers 120 to 159: representation of the last bid_

So we now have the input record for a bid as this (Q7.AT43.QT3.AT63)

[ 0, 0, 0.5, -0.71, 0.43, -0.11, 0.43,
0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 2, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 2,
1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

3 times PAD_START just means that the player is first hand to bid.

**Reply by ThorvaldAagaard** [reply, marked as answer]
The two leading zeroes are vulnaribility of NS and EW respectively but I have since added system NS and System EW as the two more information input.
I will update the documentation, thx 
I am adding a 4th bid, and it seems to work and improving the bidding. Exciting :-)
